src/python/pants/backend/python/typecheck/pyright/rules.py

```diff
@@ -35,23 +46,83 @@ class PyrightRequest(CheckRequest):
 
 
 @rule(desc="Typecheck using Pyright", level=LogLevel.DEBUG)
-async def pyright_typecheck(request: PyrightRequest, pyright: Pyright) -> CheckResults:
+async def pyright_typecheck(
+    request: PyrightRequest,
+    pyright: Pyright,
+    pex_env: PexEnvironment,
+) -> CheckResults:
     if pyright.skip:
         return CheckResults([], checker_name=request.tool_name)
 
+    coarsened_targets = await Get(
+        CoarsenedTargets,
+        CoarsenedTargetsRequest(field_set.address for field_set in request.field_sets),
+    )
+
+    coarsened_sources = await Get(
+        PythonSourceFiles, PythonSourceFilesRequest(coarsened_targets.closure())
+    )
+
     source_files = await Get(
         SourceFiles, SourceFilesRequest([field_set.sources for field_set in request.field_sets])
     )
 
+    # See `requirements_venv_pex` for how this will get wrapped in a `VenvPex`.
+    requirements_pex = await Get(
+        Pex,
+        RequirementsPexRequest(
+            (fs.address for fs in request.field_sets),
+            # TODO: Setup the correct interpreter constraints after partitioning
+            # hardcoded_interpreter_constraints=partition.interpreter_constraints,
+        ),
+    )
+
+    requirements_venv_pex = await Get(
+        VenvPex,
+        PexRequest(
+            output_filename="requirements_venv.pex",
+            internal_only=True,
+            pex_path=[requirements_pex],
+            # TODO: Setup the correct interpreter constraints after partitioning
+            interpreter_constraints=InterpreterConstraints(["==3.9.15"]),
```

Comment:
Hardcoded for SJ's machine for testing

Reply:
@Eric-Arellano Struggling a bit with the various `interpreter_constraints` to fill in here, while still working. I've taken the `example-python` repo and added a `numpy` dependency to my requirements.txt test whether I've set up everything correctly (which I haven't). On my system, I have Python3.9 and Python3.10 installed - under default constraints (>=3.7,<4), the `numpy` that gets installed only works under Py3.9 wheels. 

If I remove the `3.9.15` above, then `3.10` gets picked up, and `numpy` doesn't install. 

In `pants.toml`, I specify `interpreter_constraints = ["==3.9.*"]` and then generate lockfiles - cool beans.

Now, in my `pyright` subsystem, I create a `default_interpreter_constraints=[">=3.7,<4"]`.

Then, I also have a `field_set` `InterpreterConstraintsField`. So I have 3 possible sets of interpreter constraints, and to my surprise, it looks like the global `pants.toml` one doesn't propagate to each target, as each `field_set` is coming up `interpreter_constraints=None`.

So... In summary... We have `pants.toml`, target, subsystem, and lockfile (which should match pants.toml) interpreter_constraints. What's the correct way to reconcile all of these, so that the correct one is used throughout the `check` process?